text/0119-enhance-slow-score-store-scheduler.md

```diff
@@ -0,0 +1,168 @@
+1. **RatioMaxThresh** is a maximum tolerance threshold for the timeout ratio. When it exceeds this value, we consider the store unavailable.
+    - For disk io slow store, RatioMaxThresh was set to 0.1
+    - For network slow store, RatioMaxThresh will be set to 1.0
+2. Next, the timeout rate is normalized.
+    $$normalizedTimeoutRatio = min(TimeoutRatio, RatioMaxThresh) / RatioMaxThresh$$
+3. Finally, `NetworkSlowScore` is enlarged according to the following formula
+    $$NetworkSlowScore = min(100, NetworkSlowScore * (1 + normalizedTimeoutRatio))$$
+
+#### Score decay
+
+100 to 1 is a linear decrease over time. If the store's network is restored, we assume it will complete this process within **MinTTR**. Therefore, the formula for calculating the NetworkSlowScore is 
+$$NetworkSlowScore = NetworkSlowScore - (100 * (InspectInterval / MinTTR))$$
+
+- For disk io slow store, MinTTR is set to 5 min.
+- For network slow store, MinTTR will be set to 10 min.
+
+### How to collect TimeoutRatio
+
+There are two ways to collect TimeoutRatio:
+
+- Create a PD <-> TiKV health check mechanism. Directly use the information obtained by PD to determine the network status of TiKV. 
+- Establish TiKV <-> TiKV health check mechanism. Each TiKV still uploads the health information obtained by the health check through store heartbeat.
+
+#### PD <-> TiKV health check
+
+PD <-> TiKV health check is equivalent to a higher-frequency store heartbeat. However, it is only responsible for health checks and does not mix other functions.
+
+PD <-> TiKV health check is a very simple framework. However, it can't handle this situation that the problematic tikv is only experiencing network problems with other tikv nodes, but the network is normal with the pd nodes.
+
+#### TiKV <-> TiKV health check
+
+When a tikv has a network problem, other tikvs will also experience request timeouts when accessing the problematic tikv. In this case, the slow score of the normal tikv may also increase, and if it lasts for a long time, it may even reach 100, triggering the slow store mechanism.
+
+To avoid this problem, we can send the calculated network slow score and its corresponding store_id to PD. If the `NetworkSlowScore` of a node equals to 1, it is considered that the network of the node is normal, and not be sent to PD to reduce the traffic packet size.
+
+#### Summary
+
+~~Since the first method can quickly and effectively solve most scenarios, we will deliver the first method now and then complete the second method.~~
+
+In the latest implementation, we completed the second method.
+
+### HealthChecker
+
+It introduces a dedicated multi-threaded `HealthChecker` that probes every store once per `InspectInterval`. So each time SlowScore is computed it can directly read the latest network latency information from the `HealthChecker`.
```

Comment:
```suggestion
It introduces a dedicated multi-threaded `HealthChecker` that probes every store once per `InspectInterval`. Therefore, each time the SlowScore is computed, it can directly fetch the latest network latency information from the `HealthChecker` .
```